tests/test_root_commands.py

```diff
@@ -131,3 +131,19 @@ def complete(partialPath):
     assert len(complete('monitors.focus.')) >= 8
     assert complete('t') == ['tags.', 'theme.', 'tmp.']
     assert complete('') == [l + '.' for l in hlwm.list_children_via_attr('')]
+
+@pytest.mark.parametrize('attrtype', ['int', 'bool', 'string', 'color', 'uint'])
+@pytest.mark.parametrize('name', ['my_test', 'my_foo'])
+@pytest.mark.parametrize('object_path', ['', 'clients', 'theme.tiling.active'])
+def test_new_attr(hlwm, attrtype, name, object_path):
+    path = (object_path + '.' + name).lstrip('.')
+
+    hlwm.call(['new_attr', attrtype, path])
+
+    hlwm.get_attr(path)
+
+    hlwm.call(['remove_attr', path])
```

Comment:
So this test is actually verifying parts of the behavior of three different things, `new_attr`, `get_attr` and `remove_attr`, right? It's obviously not an overly large test case, but the way it is now, I think it's just too easy to miss some interesting corner cases for the individual commands:
 * `new_attr`: What if the name isn't valid? (I already said that above)
 * `get_attr`: What if the attribute doesn't exist? (a test for that is already "hidden" in here)
 * `remove_attr`: What if the attribute doesn't exist? What if you try to remove a non-user-defined attribute?
 * And another one, that hasn't appeared yet `set_attr`: Can the custom attribute be written at all?

So, from test design point of view, I think it would make a lot of sense to start with a `test_new_attr_` prefix(!), think of which specific behavior could be tested there, and then continue this with the other two commands. I'd even consider putting this into a separate `test_attr.py`.

Sidenote: While writing this, I already discovered that the error message for attempting to remove a non-custom attribute is rather not helpful and should probably be improved (fwiw, it's worse in master). Things like that are one of the reasons I see value in these stupidly small test cases. It's not _just_ me being overly pedantic about simple behavior that is unlikely to actually break by accident.

Reply:
I think this only tests `new_attr` and `remove_attr`. And I don't claim to fully test them. Actually I only wanted to create a test case that failed before this PR and works after this PR. But I have just noticed that I messed this up (in order to check the fix of this PR, I must not remove the user-defined attribute again).

Reply:
Yeah, these kinds of tests (for the absence of a previously known bug) are sometimes at odds with how you'd put the tests if the bug had never existed. And don't get me wrong, I didn't assume any claim about the completeness of the test, it just looked to me like the foundation of a messy set of tests. Maybe mostly because of the generic name.

How about turning this into a specific "test for the absence of a memory leak with custom attributes" test and naming it accordingly? After all, having a leak due to entities created at runtime is not far-fetched at all (even if we didn't actually have this bug).

Reply:
Yes, I've clarified the name of the test case and fixed the testcase to test the cleanup during hlwm shutdown.